### ISSUE TITLE
How to update custom pivot database table with new columns

### ISSUE THREAD

**Issue by loetie** [zotonic/zotonic#1206]
Currently the documentation shows how to add a custom pivot, which is great!
https://github.com/zotonic/zotonic/blob/master/doc/cookbook/custom-pivot.rst

However it does not show how to update the database table when you add new columns to the custom pivot table. We tried to build the site and then restarting the site or server, but that does not seem to do the trick.

We had to remove the table from the database within postgress and then run the init command to get the new structure in place.

Maybe it would be a good idea to:
A - Add it to the documentation
B - Make it easier to update the database schema when adding new columns to the pivot table.

**Comment by mworrell** [comment]
We generally:

 * Increment the module's schema version on pivot table changes
 * Add some SQL to check and upgrade the table on `manage_schema/2` calls.

**Comment by loetie** [comment]
Does seem little bit double because you also have to do the "z_pivot_rsc:define_custom_pivot" and then also write the SQL that normally would be generated in the function above.

Wouldn't it be better to alter the table automaticly when restarting the site and then run the reindex function? 

So the extend the current create table check when restarting the site with an update option?

**Comment by mworrell** [comment]
Re-running pivoting is a bit over the top ... especially for larger systems.

Right now we don't have anything planned to do automatic schema upgrades based on the definitions.
(To do that well is also a non trivial problem)

**Comment by ddeboer** [comment]
I agree with @loetie that the developer experience with defining pivots can be improved.

> (To do that well is also a non trivial problem)

True, but what if we keep it simple and change `define_custom_pivot` to take an argument like this:

```erlang
z_pivot_rsc:define_custom_pivot(
    my_pivot_name,
    [
        {age, <<"integer">>"},
        {name, <<"varchar(100">>}
    ]
```

It wouldn’t be too hard to cross-reference the tuple keys with database column names:

1. Iterate over tuple keys (`age`, `name`) and check whether there are corresponding database columns and indices.
2. Iterate over database columns and remove any that we don’t kind in the list of tuples.

**Comment by mworrell** [comment]
You can use `#column{}` definitions from z_db for the column definitions.

**Comment by mworrell** [comment]
Even better, just define in the "custom pivot" definition also an optional callback for mapping the resource properties before the core custom pivot routines fetch the properties. This should, of course, be optional.

**Comment by ddeboer** [comment]
Agreed on using `#column{}` here.

> Even better, just define in the "custom pivot" definition also an optional callback for mapping the resource properties before the core custom pivot routines fetch the properties. This should, of course, be optional.

I’m not sure I understand. Mapping what to what?

**Comment by mworrell** [comment]
Mapping the resource properties used by the system pivoter to the custom-pivot properties.

BTW Ideally I would change all pivots to the same "custom" pivot method, so that we can make them pluggable and get rid of a big chunk of indices on the rsc table.

**Comment by ddeboer** [comment]
You mean Zotonic’s built in pivots (as opposed to custom pivots defined by developers)?

**Comment by mworrell** [comment]
Yes, the ones in z_pivot_rsc

**Comment by ddeboer** [comment]
Right, so having the same mechanism for custom pivots and the built-in ones (`rsc_pivot_tsv_key`, `rsc_pivot_surname_key` and the like).

**Comment by ddeboer** [comment]
Resolved by #1735.
